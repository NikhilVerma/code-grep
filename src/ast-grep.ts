/* eslint-disable no-console */
/* eslint-env node */

/**
 * This file is a tool to "grep" for code by using Abstract syntax trees
 * It allows developers to do refactorings by searching for code patterns
 * which would be hard to express using regular expressions
 *
 * - Member expressions create({ something }) vs xxx.create({ something })
 * - Detect file contents (to use TS or Flow etc)
 * - Support block statements - { onOpen_: null }
 */
import { parse, ParserPlugin } from '@babel/parser';
import traverse, { Node, NodePath, TraverseOptions } from '@babel/traverse';
import { Statement } from '@babel/types';
import chalk from 'chalk';
import fs from 'fs';
import glob from 'glob';
import path from 'path';
import { argv } from 'yargs';

const pattern = argv._[0];

const SKIP_MASK_KEYS = ['start', 'end', 'loc', 'selfClosing', 'extra'] as const;

function createMask<T extends Record<string, any> | Record<string, any>[]>(node: T): T {
    if (node instanceof Array) {
        return node.map(createMask);
    } else if (node instanceof Object) {
        return (Object.keys(node) as Array<keyof T>).reduce((out, key) => {
            if (SKIP_MASK_KEYS.indexOf(key) === -1 && !isEmpty(node[key])) {
                out[key] = createMask(node[key]);
            }
            return out;
        }, {} as T);
    } else {
        return node;
    }
}

const expressionMask = getMappedMask(pattern);

function getMappedMask(code: string): Statement[] {
    return createMask(parseCode(`(${code})`).program.body).map((mask) => {
        // A lot of code is parsed as an expression statement while the intention of the user
        // is to search for the expression inside the statement
        if (mask.type === 'ExpressionStatement') {
            return mask.expression;
        }

        return mask;
    });
}

console.log(JSON.stringify(expressionMask, null, 4));

//@todo - compile glob from supported languages
const globPattern = (argv.glob as string) || '**/*.{js,jsx,ts,tsx}';

const cwd = argv.cwd ? path.resolve(argv.cwd as string) : process.cwd();

glob.sync(globPattern, {
    absolute: true,
    dot: true,
    cwd,
    ignore: [
        //@todo - use gitignore files - https://www.npmjs.com/package/ignore
        // exclude
        '**/node_modules/**',
        '**/target/**',
        '**/js/cosmos/**',
    ],
}).forEach(astFinder);

function astFinder(fileName: string) {
    const code = fs.readFileSync(fileName, 'utf8');

    try {
        const ast = parseCode(code, fileName);

        const visitor = getVisitorFromMask(expressionMask[0], (nodePath) => {
            console.log(fileName, nodePath.toString());
        });

        traverse(ast, visitor);
    } catch (e) {
        console.log(chalk.bgRed(`Failed to parse ${fileName}. Call with --debug to see stack`));

        if (argv.debug) {
            console.error(e);
        }
    }
}

function getVisitorFromMask(mask: Node, maskFound: (nodePath: NodePath<Node>) => void) {
    const type = mask.type;

    return {
        [type](nodePath: NodePath<Node>) {
            if (matches(nodePath.node, mask)) {
                maskFound(nodePath);
            }
        },
    } as TraverseOptions;
}

function isObject(value: any): value is object {
    return Object.prototype.toString.call(value) === '[object Object]';
}

function isArray(value: any): value is [] {
    return Object.prototype.toString.call(value) === '[object Array]';
}

function matches(value: Node, mask: any): boolean {
    if (Object.prototype.toString.call(mask) === '[object RegExp]') {
        return mask.test(value);
    } else if (isObject(mask)) {
        return (
            isObject(value) &&
            //@ts-ignore
            !Object.entries(mask).some(([key, maskValue]) => !matches(value[key], maskValue))
        );
    } else if (isArray(mask) && isArray(value)) {
        return mask.every((maskItem) => value.some((valueItem) => matches(valueItem, maskItem)));
    }

    return value === mask;
}

// eslint-disable-next-line complexity
function isEmpty(val: any) {
    return (
        val === undefined ||
        val === null ||
        val === '___' ||
        (val instanceof Array && val.length === 0) ||
        (val instanceof Object && Object.keys(val).length === 0)
    );
}

function parseCode(code: string, fileName?: string) {
    //@todo - better TS detection
    const isTypescript = fileName?.endsWith('tsx') || fileName?.endsWith('ts');

    return parse(code, {
        allowImportExportEverywhere: true,
        plugins: [
            'asyncGenerators',
            'bigInt',
            'classProperties',
            'dynamicImport',
            'exportDefaultFrom',
            'functionBind',
            'functionSent',
            'importMeta',
            'jsx',
            'nullishCoalescingOperator',
            'numericSeparator',
            'objectRestSpread',
            'optionalCatchBinding',
            'optionalChaining',
            'throwExpressions',

            // Shaves a second off parsing for 2k+ files
            isTypescript ? 'typescript' : false,
            isTypescript
                ? [
                      'decorators',
                      {
                          decoratorsBeforeExport: true,
                      },
                  ]
                : false,

            // 'classPrivateMethods',
            // 'classPrivateProperties',
            // 'doExpressions',
            // 'estree',
            // 'exportNamespaceFrom',
            // 'pipelineOperator',
        ].filter(Boolean) as ParserPlugin[],
    });
}
