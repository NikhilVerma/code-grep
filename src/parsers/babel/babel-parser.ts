import { parse, ParserPlugin } from '@babel/parser';

export function babelParser(code: string, fileName?: string) {
    //@todo - better TS detection
    const isTypescript = fileName?.endsWith('tsx') || fileName?.endsWith('ts');

    return parse(code, {
        allowImportExportEverywhere: true,
        plugins: [
            'asyncGenerators',
            'bigInt',
            'classProperties',
            'dynamicImport',
            'exportDefaultFrom',
            'functionBind',
            'functionSent',
            'importMeta',
            'jsx',
            'nullishCoalescingOperator',
            'numericSeparator',
            'objectRestSpread',
            'optionalCatchBinding',
            'optionalChaining',
            'throwExpressions',

            // Shaves a second off parsing for 2k+ files
            isTypescript ? 'typescript' : false,
            isTypescript
                ? [
                      'decorators',
                      {
                          decoratorsBeforeExport: true,
                      },
                  ]
                : false,

            // 'classPrivateMethods',
            // 'classPrivateProperties',
            // 'doExpressions',
            // 'estree',
            // 'exportNamespaceFrom',
            // 'pipelineOperator',
        ].filter(Boolean) as ParserPlugin[],
    });
}
