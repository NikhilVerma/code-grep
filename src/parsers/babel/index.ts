export { babelParser as default } from './babel-parser';
export { CodeMask, createCodeMask } from './create-mask';
