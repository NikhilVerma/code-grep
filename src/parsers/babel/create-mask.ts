import { babelParser } from './babel-parser';
import { Statement, Expression } from '@babel/types';
import { Node } from '@babel/traverse';

const SKIP_MASK_KEYS = ['start', 'end', 'loc', 'selfClosing', 'extra'];

export type CodeMask = ReturnType<typeof createCodeMask>[number];

export function createCodeMask(code: string) {
    return createMask(getParsedCode(code).program.body).map(expandMask).flat();
}

type NodeType = Node | Statement | Expression;

function expandMask(mask: Partial<NodeType>): Array<Partial<NodeType>> {
    // A lot of code is parsed as an expression statement while the intention of the user
    // is to search for the expression inside the statement
    if (mask.type === 'ExpressionStatement' && mask.expression) {
        return [mask, ...expandMask(mask.expression)];
    }

    // If a user enters "<A /><B />" as their pattern, it's not valid JSX
    // so we wrap it with fragments, but the user could be looking to match neighbouring elements
    if (mask.type === 'JSXFragment') {
        return [
            { type: 'JSXFragment', children: mask.children },
            { type: 'JSXElement', children: mask.children },
        ];
    }

    //@todo - Add support for AssignmentExpression (a = 5) > VariableDeclarator (const a = 5)

    return [mask];
}

function getParsedCode(code: string) {
    try {
        return babelParser(`(${code})`);
    } catch (err) {
        // Try to fix user mistakes

        // Invalid JSX
        if (err.message.includes('Adjacent JSX elements must be wrapped in an enclosing tag')) {
            return babelParser(`<>${code}</>`);
        }

        // Self own, don't convert this code into an expression
        if (
            err.message.includes('import can only be used') ||
            err.message.includes('Unexpected token')
        ) {
            return babelParser(code);
        }

        throw err;
    }
}

function createMask(node: NodeType[]): Partial<NodeType>[];
function createMask(node: NodeType): Partial<NodeType>;

function createMask(node: any): any {
    if (node instanceof Array) {
        return node.map(createMask);
    } else if (node instanceof Object) {
        const nodeKeys: Array<keyof typeof node> = Object.keys(node);
        const output: Partial<typeof node> = {};

        return nodeKeys.reduce((out, key) => {
            key = String(key);
            if (SKIP_MASK_KEYS.indexOf(key) === -1 && !isEmpty(node[key])) {
                out[String(key)] = createMask(node[key]);
            }
            return out;
        }, output);
    } else {
        return node;
    }
}

// eslint-disable-next-line complexity
function isEmpty(val: any) {
    return (
        val === undefined ||
        val === null ||
        val === '___' ||
        (val instanceof Array && val.length === 0) ||
        (val instanceof Object && Object.keys(val).length === 0)
    );
}
