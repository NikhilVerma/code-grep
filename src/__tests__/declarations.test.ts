import { codeSearch } from 'ast-grep';
import { createCodeMask } from 'parsers/babel';

test('ImportDeclaration', () => {
    const mask = createCodeMask(`import abc from 'xxx';`);
    let matches: string[] = [];

    codeSearch(
        `
        import abc, { something } from 'xxx';
        import { something2 } from 'xxx';
        import 'xxx';
        import 'abc';
`,
        mask,
        function (nodePath) {
            matches.push(nodePath.toString());
        }
    );

    expect(matches).toMatchSnapshot();
});

test('ImportDeclarationWildcard', () => {
    const mask = createCodeMask(`import ___ from 'xxx';`);
    let matches: string[] = [];

    console.log(JSON.stringify(mask, null, 4));

    codeSearch(
        `
        import abc, { something } from 'xxx';
        import bcd from 'xxx';
        import bcd123 from '123';
`,
        mask,
        function (nodePath) {
            matches.push(nodePath.toString());
        }
    );

    expect(matches).toMatchSnapshot();
});
