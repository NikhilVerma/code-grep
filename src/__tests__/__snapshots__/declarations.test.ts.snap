// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportDeclaration 1`] = `
Array [
  "import abc, { something } from 'xxx';",
]
`;

exports[`ImportDeclarationWildcard 1`] = `
Array [
  "import abc, { something } from 'xxx';",
  "import bcd from 'xxx';",
]
`;
